src/java/org/apache/cassandra/db/filter/ColumnFilter.java

```diff
@@ -502,149 +561,473 @@ public ColumnFilter build()
-            // filters when the cluster contains 3.4- nodes. We do that by forcibly setting queried to null.
-            //
-            // see CASSANDRA-10657, CASSANDRA-15833, CASSANDRA-16415
-            return new ColumnFilter(isFetchAll, isFetchAll && shouldFetchAllStatics(), metadata, isFetchAll && shouldQueriedBeNull() ? null : queried, s);
+            return s;
         }
     }
 
-    @Override
-    public boolean equals(Object other)
+    /**
+     * {@code ColumnFilter} sub-class for wildcard queries.
+     *
+     * <p>The class  does not rely on TableMetadata and expect a fix set of columns to prevent issues
+     * with Schema race propagation. See CASSANDRA-15899.</p>
+     */
+    public static class WildCardColumnFilter extends ColumnFilter
     {
-        if (other == this)
+        /**
+         * The queried and fetched columns.
+         */
+        private final RegularAndStaticColumns fetchedAndQueried;
+
+        /**
+         * Creates a {@code ColumnFilter} for wildcard queries.
+         *
+         * <p>The class  does not rely on TableMetadata and expect a fix set of columns to prevent issues
+         * with Schema race propagation. See CASSANDRA-15899.</p>
+         *
+         * @param fetchedAndQueried the fetched and queried columns
+         */
+        private WildCardColumnFilter(RegularAndStaticColumns fetchedAndQueried)
+        {
+            this.fetchedAndQueried = fetchedAndQueried;
+        }
+
+        @Override
+        public RegularAndStaticColumns fetchedColumns()
+        {
+            return fetchedAndQueried;
+        }
+
+        @Override
+        public RegularAndStaticColumns queriedColumns()
+        {
+            return fetchedAndQueried;
+        }
+
+        @Override
+        public boolean fetchesAllColumns(boolean isStatic)
+        {
             return true;
+        }
 
-        if (!(other instanceof ColumnFilter))
-            return false;
+        @Override
+        public boolean allFetchedColumnsAreQueried()
+        {
+            return true;
+        }
 
-        ColumnFilter otherCf = (ColumnFilter) other;
+        @Override
+        public boolean fetches(ColumnMetadata column)
+        {
+            return true;
+        }
 
-        return otherCf.fetchAllRegulars == this.fetchAllRegulars &&
-               otherCf.fetchAllStatics == this.fetchAllStatics &&
-               Objects.equals(otherCf.fetched, this.fetched) &&
-               Objects.equals(otherCf.queried, this.queried) &&
-               Objects.equals(otherCf.subSelections, this.subSelections);
-    }
+        @Override
+        public boolean fetchedColumnIsQueried(ColumnMetadata column)
+        {
+            return true;
+        }
 
-    @Override
-    public String toString()
-    {
-        String prefix = "";
+        @Override
+        public boolean fetchedCellIsQueried(ColumnMetadata column, CellPath path)
+        {
+            return true;
```

Comment:
I was going to ask about whether it works correctly and I noticed that you removed the relevant test case - why?

Reply:
similar question for tester

Reply:
I believe that the existing logic was due to the complexity of the class rather than to a realistic usecase.
True wildcard queries do not allow sub-selections (as it will not be a wildcard query otherwise).
The issue was with the fact that we used also wildcard queries for pre 3.4 nodes. I checked those versions and they rely on `isFetchAll` to avoid unecessary work. As the wildcard query will result on those version on `isFetchAll=true` the sub-selection will never be checked and that part of the logic can be removed.
Of course, I might have missed something. :-)

Reply:
Well, I only know that test was identical to what we test in 3.0 (https://github.com/apache/cassandra/blob/0e1f217079fa01366d9c8d758498016300467d31/test/unit/org/apache/cassandra/db/filter/ColumnFilterTest.java#L333). That is, it is possible to build a column filter in 3.0 with fetch-all, which has specified subselections and which can distinguish between fetched and queried for the selected cells. In fact, when I was writing those tests, I implemented them on 3.0 and copied to 3.11, checked if they work the same way assuming the cluster is on 3.0, then copied all the tests to 4.0 and did similar thing. So anything you change only in 4.0 for 3.0-tests, will be a behaviour different to what we have in 3.0 with the same filter.

Reply:
In the 3.0 version the only use case where sub-selections are used is for slice queries on super column and the code always use a `selectionBuilder` for those scenarios. By consequence, the test use case is not a true use case in 3.0.  
Moreover, sub selection through CQL queries is only possible since 4.0 (see CASSANDRA-7396) and Thrift support has been removed in 4.0 so mix-cluster queries (3.0-4.0 or 3.11-4.0)  involving sub-selections should never occurs.